client/finality-grandpa/rpc/src/lib.rs

```diff
@@ -0,0 +1,190 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! RPC API for GRANDPA.
+
+#![deny(missing_docs)]
+
+use finality_grandpa::BlockNumberOps;
+use sc_finality_grandpa::{voter, AuthorityId, SharedAuthoritySet, SharedVoterState};
+
+use futures::{FutureExt, TryFutureExt};
+use jsonrpc_derive::rpc;
+use serde::{Deserialize, Serialize};
+use std::{collections::HashSet, fmt::Debug};
+
+const NOT_READY_ERROR_CODE: i64 = 1;
+
+type FutureResult<T> =
+	Box<dyn jsonrpc_core::futures::Future<Item = T, Error = jsonrpc_core::Error> + Send>;
+
+#[derive(derive_more::Display, derive_more::From)]
+enum Error {
+	#[display(fmt = "GRANDPA RPC endpoint not ready")]
+	EndpointNotReady,
+}
+
+impl From<Error> for jsonrpc_core::Error {
+	fn from(error: Error) -> Self {
+		jsonrpc_core::Error {
+			message: format!("{}", error).into(),
+			code: jsonrpc_core::ErrorCode::ServerError(NOT_READY_ERROR_CODE),
+			data: None,
+		}
+	}
+}
+
+/// Provides RPC methods for interacting with GRANDPA.
+#[rpc]
+pub trait GrandpaApi {
+	/// Returns the state of the current best round state as well as the
+	/// ongoining backgrounds rounds.
+	#[rpc(name = "grandpa_roundState")]
+	fn round_state(&self) -> FutureResult<ReportedRoundStates>;
+}
+
+/// Implements the GrandpaApi RPC trait for interacting with GRANDPA.
+pub struct GrandpaRpcHandler<Hash, Block> {
+	shared_voter_state: SharedVoterState,
+	shared_authority_set: SharedAuthoritySet<Hash, Block>,
+}
+
+impl<Hash, Block> GrandpaRpcHandler<Hash, Block> {
+	/// Creates a new GrandpaRpcHander instance.
+	pub fn new(
+		shared_voter_state: SharedVoterState,
+		shared_authority_set: SharedAuthoritySet<Hash, Block>,
+	) -> Self {
+		Self {
+			shared_voter_state,
+			shared_authority_set,
+		}
+	}
+}
+
+#[derive(Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct Prevotes {
+	current_weight: u64,
```

Comment:
Is it possible to lower this down to `u32`? Problem with `u64` in JS is that it might overflow the builtin types, so we usually encode numbers that can possibly saturated `u64` space as hex to avoid client errors.
If this is not expected to reach `u64` then maybe it's possible to make the type smaller here?

Reply:
Ah good point, will lower to u32!

Reply:
The current implementation supports weighted validator sets, but in practice we always set all validators to have weight 1 since we use phragmen elections for staking. Therefore the maximum weight will be equal to the validator set size, so yeah `u32` should be more than enough here.